### ISSUE TITLE
Implement (main) missing schedulers

### ISSUE THREAD

**Issue by mspronesti** [LaurentMazare/diffusers-rs#23]
This issue aims at tracking the main schedulers missing in this Rust implementation:

- [x] Euler Discrete ([Python version](https://github.com/huggingface/diffusers/blob/main/src/diffusers/schedulers/scheduling_euler_discrete.py))
- [x] Euler Ancestral Discrete ([Python version](https://github.com/huggingface/diffusers/blob/main/src/diffusers/schedulers/scheduling_euler_ancestral_discrete.py))
- [x] LMS Discrete ([Python version](https://github.com/huggingface/diffusers/blob/main/src/diffusers/schedulers/scheduling_lms_discrete.py))
- [x] Heun Discrete ([Python version](https://github.com/huggingface/diffusers/blob/main/src/diffusers/schedulers/scheduling_heun_discrete.py))
- [x] K-DPM-2 Discrete ([Python Version](https://github.com/huggingface/diffusers/blob/main/src/diffusers/schedulers/scheduling_k_dpm_2_discrete.py))
- [x] K-DPM-2 Ancestral Discrete ([Python Version](https://github.com/huggingface/diffusers/blob/main/src/diffusers/schedulers/scheduling_k_dpm_2_ancestral_discrete.py))
- [x] PNDM [^1] ([Python version](https://github.com/huggingface/diffusers/blob/main/src/diffusers/schedulers/scheduling_pndm.py))

[^1]: only the relevant part for stable diffusion, i e. skipping Runge-Kutta steps.

I will be working on the first four in the following days/weeks, if you agree. Just wanted to discuss with you the following aspect: they all require a linear interpolator. The python version uses `np.interpolate(.)`. Do you prefer to introduce an extra dependency using a third-party interpolation lib or to implement "manually" a linear interpolator and to place it, for instance, at package level in `mod.rs` ? 

Feel free to update this issue adding any other scheduler you want include😄.

**Comment by rockerBOO** [comment]
Is DPM-Solver++(2M) (ancestral) possible (already)?

**Comment by mspronesti** [comment]
@rockerBOO Do you mean [this](https://github.com/huggingface/diffusers/blob/main/src/diffusers/schedulers/scheduling_k_dpm_2_ancestral_discrete.py) one ?
Anyway right now, this implementation only supports ddim, ddpm and dpm solver multistep.

**Comment by rockerBOO** [comment]
Specifically, this one in k-diffusion [sample_dpmpp_2s_ancestral](https://github.com/crowsonkb/k-diffusion/blob/master/k_diffusion/sampling.py#L509-L539). I think the DPM Solver Multistep would be accurate, but not sure. 

Also in the AUTOMATIC1111 repo [sampler/scheduler list](https://github.com/AUTOMATIC1111/stable-diffusion-webui/blob/master/modules/sd_samplers.py#L21-L39)

- DPM++ 2M
- DPM++ 2M Karras
- DPM++ 2S a
- DPM++ 2S a Karras

Not sure the ones in the HF diffusers library to translate the implementation over. The ones I listed are the ones I use most of the time now. I _was_ using DPM 2 Ancestral before, but the lower samples for DPM++ 2M/DPM++ 2S are nice. Again, maybe the DPM Solver multistep is what this is called? 

Also, the Karras component, is that part of these schedulers? Like this function to get the [sigmas karras](https://github.com/crowsonkb/k-diffusion/blob/master/k_diffusion/sampling.py#L17-L23). 

I'm a little green in what all this truly means, but have been trying to learn.

**Comment by LaurentMazare** [comment]
@mspronesti sounds like a great plan. Re interpolate, having a small helper function in `mod.rs` sounds fine. This could probably be done in a couple torch calls, e.g. using the basic approximation suggested in this [pytorch issue](https://github.com/pytorch/pytorch/issues/50334#issuecomment-1000917964).

**Comment by JohnAlcatraz** [comment]
> Anyway right now, this implementation only supports [ddim](https://github.com/huggingface/diffusers/blob/main/src/diffusers/schedulers/scheduling_ddim.py), [ddpm](https://github.com/huggingface/diffusers/blob/main/src/diffusers/schedulers/scheduling_ddpm.py) and [dpm solver multistep](https://github.com/huggingface/diffusers/blob/main/src/diffusers/schedulers/scheduling_dpmsolver_multistep.py). Do you want me to add DPM 2 Ancestral Discrete to the tasklist? 😄

How can I specify which of those schedulers to use at the moment? I can't seem to find a flag like `--scheduler`.

**Comment by mspronesti** [comment]
> @mspronesti sounds like a great plan. Re interpolate, having a small helper function in `mod.rs` sounds fine. This could probably be done in a couple torch calls, e.g. using the basic approximation suggested in this [pytorch issue](https://github.com/pytorch/pytorch/issues/50334#issuecomment-1000917964).

@LaurentMazare thanks for the reference. Here's a sketch of the implementation. Seems to be working however I couldn't find a way to avoid appending tensors to a `vec![]` and then stacking them (which seems not the most efficient thing to do). Mainly, `tch::Tensor::ge` has a different behavior than what I'm used to in `pytorch` so I had to "re-write" that part

```rust
    fn interp(x: &[f64], xp: Tensor, yp: Tensor) -> Tensor {
        // (fp[1:] - fp[:-1]) / (xp[1:] - xp[:-1])
        let m = (yp.slice(0, 1, None, 1) - yp.slice(0, 0, -1, 1)) /
            (xp.slice(0, 1, None, 1) - xp.slice(0, 0, -1, 1));

        // fp[:-1] - (m * xp[:-1])
        let b = yp.slice(0, 0, -1, 1) - (&m * xp.slice(0, 0, -1, 1));

        let mut tensors = vec![];
        for &t in x.iter() {
            tensors.push(xp.le(t).sum(Kind::Int64) - 1);
        }
        let indices = Tensor::stack(&tensors, 0)
            .clamp(0, m.size1().unwrap() - 1);

        m.take(&indices) * Tensor::of_slice(&x) + b.take(&indices)
    }
```

Since you're the author of `tch-rs`, I was wondering if you have a smarter idea on how to port this part 

```py
    indicies = torch.sum(torch.ge(x[:, None], xp[None, :]), 1) - 1
    indicies = torch.clamp(indicies, 0, len(m) - 1)
```

which corresponds - in my rust implementation - to 

```rust
        let mut tensors = vec![];
        for &t in x.iter() {
            tensors.push(xp.le(t).sum(Kind::Int64) - 1);
        }
        let indices = Tensor::stack(&tensors, 0)
            .clamp(0, m.size1().unwrap() - 1);
```

**Comment by mspronesti** [comment]
Hi @LaurentMazare ,
job's done for the first two. I'm now working on the LMS scheduler. It requires at some point to compute a definite integral over two timesteps. The python version from HF uses `scipy.integrate.quad`. 

What's your preference here? Can I use `rgsl` (hence, adding a new dependency) or do you prefer that I code some helper function ? Thanks 😄

**Comment by mspronesti** [comment]
All the tasks in this list have been solved! Feel free to close it or to update it if you want additional schedulers to be supported:)

**Comment by LaurentMazare** [comment]
Thanks for contributing all these!
